### PR TITLE
EVG-14608: Filter testresults by groupId

### PR DESCRIPTION
**Pull request by SupaJoon**
[EVG-14608](https://jira.mongodb.org/browse/EVG-14608)

### Description 
allow filtering testresults by groupId. this is useful for surfacing test links on the [build page](https://logkeeper.mongodb.org/build/1c451f3fc6fbc956273be52eaa33e7ec) from an EVG GQL request. the next step is to recreate the build page in spruce using the updated TaskTests gql query

### CODE REVIEW

File: rest/model/test.go [hunk 1]
```diff
@@ -89,6 +89,10 @@ func (at *APITest) BuildFromService(st interface{}) error {
 			dispString := fmt.Sprintf("/test_log/%s?text=true", *at.Logs.LogId)
 			at.Logs.RawDisplayURL = &dispString
 		}
+
+	    if v.GroupID != "" {
```
Comment (SupaJoon) [inline rest/model/test.go:93]: what is the best way to format a Go file?

Comment (khelif96) [reply rest/model/test.go:93]: Theres a built in go linter. My editor runs it on save. maybe try `make lint` https://github.com/evergreen-ci/evergreen/blob/e533ec058611329080f2df560da2dd55871c042e/makefile#L239

Comment (ablack12) [reply rest/model/test.go:93]: you can run make lint-rest-model to just run it for this package, but it doesn't format it, it's just running lint locally. I would fine a format plugin or something for your editor.

File: graphql/resolvers.go [hunk 2]
```diff
@@ -1233,6 +1233,11 @@ func (r *queryResolver) TaskTests(ctx context.Context, taskID string, execution
 		}
 	}
 
+	groupIdParam := ""
```
Comment (ablack12) [inline graphql/resolvers.go:1236]: a one line way to do this is utility.FromStringPtr

File: rest/model/test.go (hunk 1 above)
Comment (julianedwards) [reply rest/model/test.go:93]: Use [Gofmt](https://blog.golang.org/gofmt), there should be a plugin for your editor.

**Review by julianedwards — CHANGES_REQUESTED** [review]

File: graphql/resolvers.go [hunk 3]
```diff
@@ -1188,7 +1188,7 @@ func (r *queryResolver) PatchTasks(ctx context.Context, patchID string, sorts []
 	return &patchTasks, nil
 }
 
-func (r *queryResolver) TaskTests(ctx context.Context, taskID string, execution *int, sortCategory *TestSortCategory, sortDirection *SortDirection, page *int, limit *int, testName *string, statuses []string) (*TaskTestResult, error) {
+func (r *queryResolver) TaskTests(ctx context.Context, taskID string, execution *int, sortCategory *TestSortCategory, sortDirection *SortDirection, page *int, limit *int, testName *string, statuses []string, groupID *string) (*TaskTestResult, error) {
```
[the remark this answers is not in the document]
Comment (SupaJoon) [reply graphql/resolvers.go:1191]: we actually codegen this func definition from our gql schema. we may be able to clean this up by creating GQL Inputs for common params

File: rest/model/test.go (hunk 1 above)
Comment (SupaJoon) [reply rest/model/test.go]: sweet Gofmt and lint work pretty well!!

File: rest/data/test.go [hunk 4]
```diff
@@ -141,7 +141,7 @@ func (mtc *MockTestConnector) FindTestsByTaskId(taskId, testId, testName, status
 func (tc *MockTestConnector) GetTestCountByTaskIdAndFilters(taskId, testName string, statuses []string, execution int) (int, error) {
 	return 0, nil
 }
-func (tc *MockTestConnector) FindTestsByTaskIdFilterSortPaginate(taskId, filter string, statuses []string, sortBy string, sortDir, page, limit, execution int) ([]testresult.TestResult, error) {
+func (tc *MockTestConnector) FindTestsByTaskIdFilterSortPaginate(taskId, filter string, statuses []string, sortBy, groupId string, sortDir, page, limit, execution int) ([]testresult.TestResult, error) {
 	//todo
```
[the remark this answers is not in the document]
Comment (SupaJoon) [reply rest/data/test.go]: I'll remove the comment

**Review by khelif96 — CHANGES_REQUESTED** [review]

File: model/testresult/testresult.go [hunk 5]
```diff
@@ -193,17 +194,31 @@ func TestResultCount(taskIds []string, testName string, statuses []string, execu
 	return Count(q)
 }
 
+type TestResultsFilterSortPaginateOpts struct {
```
Comment (SupaJoon) [inline model/testresult/testresult.go:197]: what's the best way to make this struct available in graphql/resolvers.go?

**Comment by SupaJoon** [comment]
I'll post here when this PR is ready for another look

**Comment by SupaJoon** [comment]
this PR is ready for another look

File: graphql/util.go [hunk 6]
```diff
@@ -36,7 +36,7 @@ import (
 )
 
 // FilterSortAndPaginateCedarTestResults takes an array of CedarTestResult objects and returns a filtered sorted and paginated version of that array
-func FilterSortAndPaginateCedarTestResults(testResults []apimodels.CedarTestResult, testName string, statuses []string, sortBy string, sortDir int, pageParam int, limitParam int) ([]apimodels.CedarTestResult, int) {
+func FilterSortAndPaginateCedarTestResults(testResults []apimodels.CedarTestResult, testName string, statuses []string, sortBy, groupId string, sortDir int, pageParam int, limitParam int) ([]apimodels.CedarTestResult, int) {
```
Comment (khelif96) [inline graphql/util.go:39]: Can this also follow the opts pattern?

**Review by khelif96 — CHANGES_REQUESTED** [review]
Sorry i should have asked this in my earlier review. everything else lgtm

File: graphql/util.go [hunk 7]
```diff
@@ -36,35 +36,49 @@ import (
...
+		if opts.GroupID != "" && testResult.GroupID != opts.GroupID {
 			match = false
 		}
 		if match {
 			filteredAndSortedTestResults = append(filteredAndSortedTestResults, testResult)
 		}
 	}
-	if sortBy != "" {
-		switch sortBy {
+	isAsc := opts.SortDir == 1
```
Comment (julianedwards) [inline graphql/util.go:71]: This seems unnecessary

File: graphql/util.go [hunk 8]
```diff
@@ -36,35 +36,49 @@ import (
 )
 
 // FilterSortAndPaginateCedarTestResults takes an array of CedarTestResult objects and returns a filtered sorted and paginated version of that array
-func FilterSortAndPaginateCedarTestResults(testResults []apimodels.CedarTestResult, testName string, statuses []string, sortBy string, sortDir int, pageParam int, limitParam int) ([]apimodels.CedarTestResult, int) {
+type FilterSortAndPaginateCedarTestResultsOpts struct {
```
Comment (julianedwards) [inline graphql/util.go:39]: Add period to the documentation above.

**Review by julianedwards — CHANGES_REQUESTED** [review]

File: rest/data/test.go [hunk 9]
```diff
@@ -90,17 +66,17 @@ func (tc *DBTestConnector) FindTestsByTaskIdFilterSortPaginate(taskId, testName
...
 	if t.DisplayOnly {
-		taskIds = t.ExecutionTasks
+		taskIDs = t.ExecutionTasks
 	} else {
-		taskIds = []string{taskId}
+		taskIDs = []string{opts.TaskID}
 	}
 
-	res, err := testresult.TestResultsFilterSortPaginate(taskIds, testName, statuses, sortBy, sortDir, page, limit, execution)
+	res, err := testresult.TestResultsFilterSortPaginate(testresult.TestResultsFilterSortPaginateOpts{TestID: opts.TestID, TaskIDs: taskIDs, TestName: opts.TestName, Statuses: opts.Statuses, SortBy: opts.SortBy, GroupID: opts.GroupID, SortDir: opts.SortDir, Page: opts.Page, Limit: opts.Limit, Execution: opts.Execution})
```
Comment (julianedwards) [inline rest/data/test.go:79]: Can you break this up into separate lines?

**Review by julianedwards — CHANGES_REQUESTED** [review]

File: graphql/resolvers.go [hunk 10]
```diff
@@ -1282,12 +1282,13 @@ func (r *queryResolver) TaskTests(ctx context.Context, taskID string, execution
...
 		if baseTask != nil {
-			baseTestResults, _ := r.sc.FindTestsByTaskId(baseTask.Id, "", "", "", 0, taskExecution)
+			baseTestResults, _ := r.sc.FindTestsByTaskId(data.FindTestsByTaskIdOpts{TaskID: baseTask.Id, Execution: taskExecution})
 			for _, t := range baseTestResults {
 				baseTestStatusMap[t.TestFile] = t.Status
 			}
 		}
-		paginatedFilteredTests, err := r.sc.FindTestsByTaskIdFilterSortPaginate(taskID, testNameParam, statusesParam, sortBy, sortDir, pageParam, limitParam, taskExecution)
+
+		paginatedFilteredTests, err := r.sc.FindTestsByTaskId(data.FindTestsByTaskIdOpts{TaskID: taskID, TestName: testNameParam, Statuses: statusesParam, SortBy: sortBy, GroupID: groupIdParam, SortDir: sortDir, Limit: limitParam, Execution: taskExecution, Page: pageParam})
```
Comment (julianedwards) [inline graphql/resolvers.go:1291]: Break this up into separate lines like below on L1234

**Review by julianedwards — CHANGES_REQUESTED** [review]

File: model/testresult/testresult.go [hunk 11]
```diff
@@ -195,15 +196,38 @@ func TestResultCount(taskIds []string, testName string, statuses []string, execu
 	return Count(q)
 }
 
+// Returns tests that match exact Execution, GroupID, Statuses, TaskID, and TestName,
```
Comment (julianedwards) [inline model/testresult/testresult.go:199]: Documentation should start with the name of the struct: `TestResultsFilterSortPaginateOpts...`

**Review by julianedwards — CHANGES_REQUESTED** [review]

File: graphql/util.go (hunk 8 above)
Comment (julianedwards) [reply graphql/util.go:38]: Where did the documentation go?

File: model/testresult/testresult.go [hunk 12]
```diff
@@ -195,15 +196,36 @@ func TestResultCount(taskIds []string, testName string, statuses []string, execu
...
+// TestResultsFilterSortPaginateOpts contains filtering, sorting and pagination options for TestResults.
+type TestResultsFilterSortPaginateOpts struct {
+	Execution int
+	GroupID   string
+	Limit     int
+	Page      int
+	SortBy    string
+	SortDir   int
+	Statuses  []string
+	TaskIDs   []string // TaskIDs is the only required option.
```
Comment (julianedwards) [inline model/testresult/testresult.go:208]: Put the comment on the line above the field.

File: rest/data/interface.go [hunk 13]
```diff
@@ -29,6 +29,20 @@ import (
 	"github.com/mongodb/amboy"
 )
 
+// FindTestsByTaskIdOpts contains filtering, sorting and pagination options for TestResults.
+type FindTestsByTaskIdOpts struct {
+	Execution int    // Execution matches exact.
```
Comment (julianedwards) [inline rest/data/interface.go:34]: I think you can remove the individual field level documentation—they are pretty self-explanatory. If you keep anything, move it to the line above the field.

**Review by julianedwards — CHANGES_REQUESTED** [review]
Few small things

**Review by julianedwards — APPROVED** [review]

**Review by khelif96 — APPROVED** [review]

**Pull request merged**